src/Workspaces/Core/Portable/NamingStyles/EditorConfig/EditorConfigNamingStyleParser.cs

```diff
@@ -0,0 +1,68 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.CodeAnalysis.Diagnostics.Analyzers.NamingStyles
+{
+    internal static partial class EditorConfigNamingStyleParser
+    {
+        /// <remarks>
+        /// The dictionary we get from the VS editorconfig API uses the same dictionary object if there are no changes, so we can cache based on dictionary
+        /// </remarks>
+        private static readonly ConditionalWeakTable<IReadOnlyDictionary<string, object>, SerializableNamingStylePreferencesInfo> _cache = new ConditionalWeakTable<IReadOnlyDictionary<string, object>, SerializableNamingStylePreferencesInfo>();
+        private static readonly object _cacheLock = new object();
+
+        public static SerializableNamingStylePreferencesInfo GetNamingStylesStringFromDictionary(IReadOnlyDictionary<string, object> allRawConventions)
+        {
+            if (!_cache.TryGetValue(allRawConventions, out var value))
+            {
+                lock (_cacheLock)
+                {
+                    if (!_cache.TryGetValue(allRawConventions, out value))
+                    {
+                        value = ParseDictionary(allRawConventions);
```

Comment:
Is the speed of ParseDictionary() a potential issue here? I think it should be fast, but I might be missing something. (If so, you can do the work outside the _cacheLock and then add it to the _cache under the _cacheLock after.)